File: blue-core/src/blue/blueLive/LiveObjectBins.java
/*
 * blue - object composition environment for csound
 * Copyright (c) 2012 Steven Yi (stevenyi@gmail.com)
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published
 * by  the Free Software Foundation; either version 2 of the License or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; see the file COPYING.LIB.  If not, write to
 * the Free Software Foundation Inc., 59 Temple Place - Suite 330,
 * Boston, MA  02111-1307 USA
 */
package blue.blueLive;

import electric.xml.Element;
import electric.xml.Elements;
import java.beans.PropertyChangeListener;
import java.beans.PropertyChangeSupport;
import java.util.Map;

/**
 *
 * @author stevenyi
 */
public class LiveObjectBins {

    private LiveObject[][] liveObjectBins;

    transient PropertyChangeSupport listeners = new PropertyChangeSupport(this);

    public LiveObjectBins() {
        this(new LiveObject[1][8]);
    }

    public LiveObjectBins(LiveObject[][] liveObjectBins) {
        this.liveObjectBins = liveObjectBins;
    }

    public LiveObjectBins(LiveObjectBins bins) {
        liveObjectBins = bins.liveObjectBins.clone();
        for (int i = 0; i < liveObjectBins.length; i++) {
            for (int j = 0; j < liveObjectBins[i].length; j++) {
                if (bins.liveObjectBins[i][j] != null) {
                    liveObjectBins[i][j] = new LiveObject(bins.liveObjectBins[i][j]);
                }
            }
        }
    }

    public static LiveObjectBins loadFromXML(Element data, Map<String, Object> objRefMap) throws Exception {

        LiveObjectBins retVal;

        if (data.getAttributeValue("columns") != null) {
            int columns = Integer.parseInt(data.getAttributeValue("columns"));
            int rows = Integer.parseInt(data.getAttributeValue("rows"));
            retVal = new LiveObjectBins(new LiveObject[columns][rows]);
        } else {
            throw new Exception("LiveObjectBins could not load");
        }

        Elements nodes = data.getElements();

        int column = 0;

        while (nodes.hasMoreElements()) {

            int row = 0;

            Element node = nodes.next();
            String name = node.getName();

            if (name.equals("bin")) {

                Elements lObjNodes = node.getElements();

                while (lObjNodes.hasMoreElements()) {
                    Element lObjNode = lObjNodes.next();
                    name = lObjNode.getName();

                    if (name.equals("liveObject")) {
                        retVal.liveObjectBins[column][row] = LiveObject.loadFromXML(lObjNode, objRefMap);
                    }
                    row++;
                }
                column++;
            }
        }

        return retVal;
    }

    public Element saveAsXML(Map<Object, String> objRefMap) {
        Element retVal = new Element("liveObjectBins");

        retVal.setAttribute("columns", Integer.toString(liveObjectBins.length));
        retVal.setAttribute("rows", Integer.toString(liveObjectBins[0].length));

        for (int i = 0; i < liveObjectBins.length; i++) {

            Element bin = retVal.addElement("bin");

            for (int j = 0; j < liveObjectBins[i].length; j++) {
                LiveObject lObj = liveObjectBins[i][j];

                if (lObj == null) {
                    bin.addElement("null");
                } else {
                    bin.addElement(lObj.saveAsXML(objRefMap));
                }
            }
        }

        return retVal;
    }

    public void setLiveObject(int column, int row, LiveObject liveObject) {
        LiveObject oldObject = liveObjectBins[column][row];

        liveObjectBins[column][row] = liveObject;

        listeners.firePropertyChange("liveObject", oldObject, liveObject);
    }

    public void insertRow(int index) {
        LiveObject[][] newBins = new LiveObject[liveObjectBins.length][liveObjectBins[0].length + 1];

        if (index < 0) {
            index = 0;
        } else if (index > liveObjectBins[0].length) {
            index = liveObjectBins[0].length;
        }

        for (int i = 0; i < newBins.length; i++) {
            int readCounter = 0;

            for (int j = 0; j < newBins[i].length; j++) {
                if (j != index) {
                    newBins[i][j] = liveObjectBins[i][readCounter];
                    readCounter++;
                }
            }
        }

        int oldCount = liveObjectBins[0].length;
        liveObjectBins = newBins;

        listeners.firePropertyChange("rowCount", oldCount, newBins[0].length);
    }

    public void removeRow(int index) {
        if (index < 0 || index >= liveObjectBins[0].length || liveObjectBins[0].length <= 1) {
            return;
        }

        LiveObject[][] newBins = new LiveObject[liveObjectBins.length][liveObjectBins[0].length - 1];

        for (int i = 0; i < liveObjectBins.length; i++) {
            int writeCounter = 0;

            for (int j = 0; j < liveObjectBins[i].length; j++) {
                if (j != index) {
                    newBins[i][writeCounter] = liveObjectBins[i][j];
                    writeCounter++;
                }
            }
        }

        int oldCount = liveObjectBins[0].length;
        liveObjectBins = newBins;

        listeners.firePropertyChange("rowCount", oldCount, newBins[0].length);
    }

    public void insertColumn(int index) {

        LiveObject[][] newBins = new LiveObject[liveObjectBins.length + 1][liveObjectBins[0].length];

        if (index < 0) {
            index = 0;
        } else if (index > liveObjectBins.length) {
            index = liveObjectBins.length;
        }

        int readCounter = 0;
        boolean inserted = false;

        for (int i = 0; i < newBins.length; i++) {
            if (!inserted && readCounter == index) {
                inserted = true;
            } else {
                System.arraycopy(liveObjectBins[readCounter], 0, newBins[i], 0, liveObjectBins[0].length);
                readCounter++;
            }
        }

        int oldCount = liveObjectBins.length;
        liveObjectBins = newBins;

        listeners.firePropertyChange("columnCount", oldCount, newBins.length);
    }

    public void removeColumn(int index) {
        if (index < 0 || index >= liveObjectBins.length || liveObjectBins.length <= 1) {
            return;
        }

        LiveObject[][] newBins = new LiveObject[liveObjectBins.length - 1][liveObjectBins[0].length];

        int writeCounter = 0;
        boolean inserted = false;

        for (int i = 0; i < liveObjectBins.length; i++) {
            if (i != index) {

                System.arraycopy(liveObjectBins[i], 0, newBins[writeCounter], 0, liveObjectBins[0].length);
                writeCounter++;
            }
        }

        int oldCount = liveObjectBins.length;
        liveObjectBins = newBins;

        listeners.firePropertyChange("columnCount", oldCount, newBins.length);
    }

    public int getColumnCount() {
        return liveObjectBins.length;
    }

    public int getRowCount() {
        return liveObjectBins[0].length;
    }

    public LiveObject getLiveObject(int column, int row) {
        return liveObjectBins[column][row];
    }

    public int getColumnForObject(LiveObject liveObject) {
        if (liveObject == null) {
            return -1;
        }

        for (int i = 0; i < liveObjectBins.length; i++) {
            for (int j = 0; j < liveObjectBins[i].length; j++) {
                if (liveObjectBins[i][j] == liveObject) {
                    return i;
                }
            }
        }
        return - 1;
    }

    public int getRowForObject(LiveObject liveObject) {
        if (liveObject == null) {
            return -1;
        }

        for (int i = 0; i < liveObjectBins.length; i++) {
            for (int j = 0; j < liveObjectBins[i].length; j++) {
                if (liveObjectBins[i][j] == liveObject) {
                    return j;
                }
            }
        }
        return -1;
    }

    /* CHANGE LISTENER */
    public void addPropertyChangeListener(PropertyChangeListener cl) {
        listeners.addPropertyChangeListener(cl);
    }

    public void removePropertyChangeListener(PropertyChangeListener cl) {
        listeners.removePropertyChangeListener(cl);
    }

    public LiveObjectSet getEnabledLiveObjectSet() {

        LiveObjectSet retVal = new LiveObjectSet();

        for (int i = 0; i < liveObjectBins.length; i++) {
            for (int j = 0; j < liveObjectBins[0].length; j++) {
                LiveObject lObj = liveObjectBins[i][j];

                if (lObj != null && lObj.isEnabled()) {
                    retVal.add(lObj);
                }
            }
        }

        return retVal;
    }

    public void setEnabledFromLiveObjectSet(LiveObjectSet liveObjectSet) {

        for (int i = 0; i < liveObjectBins.length; i++) {
            for (int j = 0; j < liveObjectBins[0].length; j++) {
                LiveObject lObj = liveObjectBins[i][j];

                if (lObj != null && liveObjectSet.contains(lObj)) {
                    lObj.setEnabled(true);
                } else {
                    lObj.setEnabled(false);
                }
            }
        }

        listeners.firePropertyChange("enabledStateChanged", false, false);

    }

    public LiveObject getLiveObjectByUniqueId(String uniqueId) {

        if (uniqueId == null) {
            return null;
        }

        for (int i = 0; i < liveObjectBins.length; i++) {
            for (int j = 0; j < liveObjectBins[0].length; j++) {
                LiveObject lObj = liveObjectBins[i][j];

                if (lObj != null && uniqueId.equals(lObj.getUniqueId())) {
                    return lObj;
                }
            }
        }

        return null;
    }

}

File: packages/blue-data/src/live/live-object-bins.ts
import { Element } from '../serialization/xml-reader';
import { ObjRefSaveMap, ObjRefLoadMap } from '../serialization/obj-ref-map';
import { BlueDataObject } from '../blue-data-object';
import { LiveObject } from './live-object';

export class LiveObjectBins implements BlueDataObject {
  private _cells: Array<Array<LiveObject | null>>;

  constructor(columns = 1, rows = 8) {
    this._cells = [];
    for (let c = 0; c < columns; c++) {
      const col: Array<LiveObject | null> = [];
      for (let r = 0; r < rows; r++) {
        col.push(null);
      }
      this._cells.push(col);
    }
  }

  static fromGrid(grid: Array<Array<LiveObject | null>>): LiveObjectBins {
    const bins = Object.create(LiveObjectBins.prototype) as LiveObjectBins;
    bins._cells = grid;
    return bins;
  }

  getColumnCount(): number {
    return this._cells.length;
  }

  getRowCount(): number {
    return this._cells.length > 0 ? this._cells[0].length : 0;
  }

  getLiveObject(column: number, row: number): LiveObject | null {
    if (column < 0 || column >= this._cells.length) return null;
    if (row < 0 || row >= this._cells[column].length) return null;
    return this._cells[column][row];
  }

  setLiveObject(column: number, row: number, obj: LiveObject | null): void {
    if (column < 0 || column >= this._cells.length) return;
    if (row < 0 || row >= this._cells[column].length) return;
    this._cells[column][row] = obj;
  }

  insertRow(index: number): boolean {
    if (index < 0) index = 0;
    const rows = this.getRowCount();
    if (index > rows) index = rows;
    for (let c = 0; c < this._cells.length; c++) {
      this._cells[c].splice(index, 0, null);
    }
    return true;
  }

  removeRow(index: number): boolean {
    if (index < 0 || index >= this.getRowCount() || this.getRowCount() <= 1) return false;
    for (let c = 0; c < this._cells.length; c++) {
      this._cells[c].splice(index, 1);
    }
    return true;
  }

  insertColumn(index: number): boolean {
    if (index < 0) index = 0;
    if (index > this._cells.length) index = this._cells.length;
    const rows = this.getRowCount();
    const newCol: Array<LiveObject | null> = [];
    for (let r = 0; r < rows; r++) {
      newCol.push(null);
    }
    this._cells.splice(index, 0, newCol);
    return true;
  }

  removeColumn(index: number): boolean {
    if (index < 0 || index >= this._cells.length || this._cells.length <= 1) return false;
    this._cells.splice(index, 1);
    return true;
  }

  getEnabledLiveObjectSet(): LiveObject[] {
    const result: LiveObject[] = [];
    for (const col of this._cells) {
      for (const obj of col) {
        if (obj !== null && obj.isEnabled()) {
          result.push(obj);
        }
      }
    }
    return result;
  }

  setEnabledFromLiveObjectSet(set: LiveObject[]): boolean {
    const setIds = new Set(set.map((o) => o.getUniqueId()));
    let changed = false;
    for (const col of this._cells) {
      for (const obj of col) {
        if (obj !== null) {
          const enabled = setIds.has(obj.getUniqueId());
          if (obj.isEnabled() !== enabled) {
            obj.setEnabled(enabled);
            changed = true;
          }
        }
      }
    }
    return changed;
  }

  getLiveObjectByUniqueId(uniqueId: string | null): LiveObject | null {
    if (!uniqueId) return null;
    for (const col of this._cells) {
      for (const obj of col) {
        if (obj !== null && obj.getUniqueId() === uniqueId) {
          return obj;
        }
      }
    }
    return null;
  }

  getColumnForObject(liveObject: LiveObject | null): number {
    if (!liveObject) return -1;
    for (let c = 0; c < this._cells.length; c++) {
      for (const obj of this._cells[c]) {
        if (obj === liveObject) return c;
      }
    }
    return -1;
  }

  getRowForObject(liveObject: LiveObject | null): number {
    if (!liveObject) return -1;
    for (const col of this._cells) {
      for (let r = 0; r < col.length; r++) {
        if (col[r] === liveObject) return r;
      }
    }
    return -1;
  }

  saveAsXML(_objRefMap?: ObjRefSaveMap): Element {
    const elem = new Element('liveObjectBins');
    elem.setAttribute('columns', this._cells.length.toString());
    const rows = this.getRowCount();
    elem.setAttribute('rows', rows.toString());

    for (const col of this._cells) {
      const bin = elem.addElement('bin');
      for (const obj of col) {
        if (obj === null) {
          bin.addElement('null');
        } else {
          bin.addElement(obj.saveAsXML(_objRefMap));
        }
      }
    }

    return elem;
  }

  static loadFromXML(data: Element, objRefMap?: ObjRefLoadMap): LiveObjectBins {
    const columnsAttr = data.getAttribute('columns');
    const rowsAttr = data.getAttribute('rows');
    if (!columnsAttr || !rowsAttr) {
      throw new Error('LiveObjectBins could not load: missing columns/rows attributes');
    }
    const columns = parseInt(columnsAttr, 10);
    const rows = parseInt(rowsAttr, 10);

    const grid: Array<Array<LiveObject | null>> = [];
    for (let c = 0; c < columns; c++) {
      const col: Array<LiveObject | null> = [];
      for (let r = 0; r < rows; r++) {
        col.push(null);
      }
      grid.push(col);
    }

    const bins = LiveObjectBins.fromGrid(grid);
    let column = 0;

    const nodes = data.getElements();
    while (nodes.hasMoreElements()) {
      const node = nodes.next();
      if (node.getName() === 'bin') {
        let row = 0;
        const lObjNodes = node.getElements();
        while (lObjNodes.hasMoreElements()) {
          const lObjNode = lObjNodes.next();
          if (lObjNode.getName() === 'liveObject' && column < columns && row < rows) {
            bins._cells[column][row] = LiveObject.loadFromXML(lObjNode, objRefMap);
          }
          row++;
        }
        column++;
      }
    }

    return bins;
  }

  deepCopy(): BlueDataObject {
    const grid: Array<Array<LiveObject | null>> = [];
    for (const col of this._cells) {
      const newCol: Array<LiveObject | null> = [];
      for (const obj of col) {
        newCol.push(obj ? obj.deepCopy() as LiveObject : null);
      }
      grid.push(newCol);
    }
    return LiveObjectBins.fromGrid(grid);
  }
}
